### PR TITLE
create empty decision when no cluster selected

### PR DESCRIPTION
**Pull request by haoqing0110**
https://github.com/open-cluster-management-io/placement/issues/85

Signed-off-by: haoqing0110 <qhao@redhat.com>

### CODE REVIEW

File: pkg/controllers/scheduling/scheduling_controller.go [hunk 1]
```diff
@@ -452,7 +452,9 @@ func (c *schedulingController) bind(
 	// maxNumOfClusterDecisions.
 	decisionSlices := [][]clusterapiv1beta1.ClusterDecision{}
 	remainingDecisions := clusterDecisions
-	for index := 0; len(remainingDecisions) > 0; index++ {
+	// if clusterDecisions is empty, append empty slice to decisionSlices,
+	// and create an PlacementDecision with empty decisions in status.
+	for index := 0; len(remainingDecisions) > 0 || len(decisionSlices) == 0; index++ {
```
Comment (qiujian16) [inline pkg/controllers/scheduling/scheduling_controller.go:457]: This is confusing, can we make it more explicit

Comment (haoqing0110) [reply pkg/controllers/scheduling/scheduling_controller.go]: fixed.

File: pkg/controllers/scheduling/scheduling_controller.go [hunk 2]
```diff
@@ -464,6 +464,12 @@ func (c *schedulingController) bind(
 		}
 		decisionSlices = append(decisionSlices, decisionSlice)
 	}
+	// if decisionSlices is empty, append one empty decision slice.
+	// so that can create an PlacementDecision with empty decisions in status.
+	if len(decisionSlices) == 0 {
+		decisionSlice := []clusterapiv1beta1.ClusterDecision{}
+		decisionSlices = append(decisionSlices, decisionSlice)
```
Comment (qiujian16) [inline pkg/controllers/scheduling/scheduling_controller.go:471]: nit: decisionSlices = append(decisionSlices, []clusterapiv1beta1.ClusterDecision{})

Comment (haoqing0110) [reply pkg/controllers/scheduling/scheduling_controller.go]: fixed.

File: pkg/controllers/scheduling/scheduling_controller.go [hunk 3]
```diff
@@ -238,16 +238,8 @@ func (c *schedulingController) syncPlacement(ctx context.Context, syncCtx factor
...
-	if status.IsError() {
-		if err := c.updateStatus(ctx, placement, int32(len(scheduleResult.Decisions())), misconfiguredCondition, satisfiedCondition); err != nil {
-			return err
-		}
-		return status.AsError()
-	}
-
 	// requeue placement if requeueAfter is defined in scheduleResult
-	if syncCtx != nil && scheduleResult.RequeueAfter() != nil {
+	if syncCtx != nil && !status.IsError() && scheduleResult.RequeueAfter() != nil {
```
Comment (qiujian16) [inline pkg/controllers/scheduling/scheduling_controller.go:242]: I think we should requeue no matter there is error or not? and should we return error finally after update status?

File: pkg/controllers/scheduling/scheduling_controller.go [hunk 4]
```diff
@@ -260,7 +252,12 @@ func (c *schedulingController) syncPlacement(ctx context.Context, syncCtx factor
 	}
 
 	// update placement status if necessary to signal no bindings
-	return c.updateStatus(ctx, placement, int32(len(scheduleResult.Decisions())), misconfiguredCondition, satisfiedCondition)
+	err = c.updateStatus(ctx, placement, int32(len(scheduleResult.Decisions())), misconfiguredCondition, satisfiedCondition)
```
Comment (qiujian16) [inline pkg/controllers/scheduling/scheduling_controller.go:255]: nit

```
if err := xx ; err !=nil {
    return err
}
```

File: pkg/controllers/framework/interface.go [hunk 5]
```diff
@@ -81,6 +81,14 @@ func (s *Status) AsError() error {
 	return errors.New(s.Message())
 }
 
+// Error returns the err when Code is "Error" or "Misconfigured".
+func (s *Status) Error() error {
+	if s.IsError() {
```
Comment (qiujian16) [inline pkg/controllers/framework/interface.go:86]: hrm, we might want to refactor here. why we have both isError and asError? couldn't we just check if returned error is nil?

File: pkg/controllers/scheduling/scheduling_controller.go (hunk 4 above)
Comment (haoqing0110) [reply pkg/controllers/scheduling/scheduling_controller.go]: fixed.

File: pkg/controllers/framework/interface.go (hunk 5 above)
Comment (haoqing0110) [reply pkg/controllers/framework/interface.go]: It's for the case when `Status` is nil I think. Remove this func. And modify the NewStatus(), to ensure AsError can always return s.err when code is Error or Misconfigured.

**Pull request merged**